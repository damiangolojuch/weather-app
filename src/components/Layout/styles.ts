import styled, { css } from "styled-components";
import { Breadcrumb, Layout as AntdLayout } from "antd";
import config from "~/styles/config";
import Background from "~/components/Background";

const { Header, Content, Footer } = AntdLayout;

export const Layout = styled(AntdLayout)`
  background-color: transparent;
`;

export const BackgroundStyled = styled(Background)`
  position: fixed;
  z-index: -1;
`;

export const headerHeight = 64;
export const footerHeight = 70;

export const ContentStyled = styled(Content)<{ $isHeaderVisible: boolean }>`
  padding: 0 50px;
  ${({ $isHeaderVisible }) =>
    css`
      min-height: calc(
        100vh ${$isHeaderVisible && `- ${headerHeight}`} - ${footerHeight}px
      );
    `};

  @media (max-width: ${config.viewports.md - 1}px) {
    padding: 0 10px;
  }

  @media (min-width: ${config.viewports.xl}px) {
    margin: 0 auto;
    width: 1000px;
  }
`;

export const HeaderStyled = styled(Header)`
  height: ${headerHeight}px;
  background: transparent;

  @media (max-width: ${config.viewports.md - 1}px) {
    padding: 0 10px;
  }

  @media (min-width: ${config.viewports.xl}px) {
    margin: 0 auto;
    width: 1000px;
  }
`;

export const BreadcrumbStyled = styled(Breadcrumb)`
  margin: 16px 0;
`;

export const SiteLayoutContent = styled.div`
  padding: 24px 0;
`;

export const FooterStyled = styled(Footer)`
  height: ${footerHeight}px;
  text-align: center;
  background-color: transparent;
  color: #fff;
  text-shadow: 0 0 2px #000;
`;
