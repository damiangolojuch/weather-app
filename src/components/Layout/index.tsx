import { Breadcrumb } from "antd";
import { FC, ReactNode } from "react";
import * as S from "./styles";

interface BreadcrumbType {
  label: string;
  href: string;
}

export interface LayoutProps {
  children?: ReactNode;
  header?: ReactNode;
  breadcrumbs?: BreadcrumbType[];
}

const Layout: FC<LayoutProps> = ({ breadcrumbs, children, header }) => (
  <S.Layout>
    <S.BackgroundStyled />
    {header && <S.HeaderStyled>{header}</S.HeaderStyled>}
    <S.ContentStyled $isHeaderVisible={!!header}>
      {Array.isArray(breadcrumbs) && breadcrumbs.length > 0 && (
        <S.BreadcrumbStyled>
          {breadcrumbs.map((breadcrumb) => (
            <Breadcrumb.Item href={breadcrumb.href}>
              {breadcrumb.label}
            </Breadcrumb.Item>
          ))}
        </S.BreadcrumbStyled>
      )}
      <S.SiteLayoutContent>{children}</S.SiteLayoutContent>
    </S.ContentStyled>
    <S.FooterStyled>Â© Aplikacja pogodowa 2022</S.FooterStyled>
  </S.Layout>
);

export default Layout;
