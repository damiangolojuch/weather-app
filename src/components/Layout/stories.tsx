import { ComponentMeta, ComponentStory } from "@storybook/react";
import Layout from ".";

export default {
  title: "Layout",
  component: Layout,
  args: {
    description:
      "Displays main page layout with optional header and breadcrumbs slots.",
  },
} as ComponentMeta<typeof Layout>;

const DefaultTemplate: ComponentStory<typeof Layout> = (args) => (
  <Layout {...args} />
);

export const Default = DefaultTemplate.bind({});

Default.args = {
  children: <div>Some content</div>,
};

const WithHeaderTemplate: ComponentStory<typeof Layout> = (args) => (
  <Layout {...args} />
);

export const WithHeader = WithHeaderTemplate.bind({});

WithHeader.args = {
  children: <div>Some content</div>,
  header: <h1>Header here</h1>,
};

const WithBreadcrumbsTemplate: ComponentStory<typeof Layout> = (args) => (
  <Layout {...args} />
);

export const WithBreadcrumbs = WithBreadcrumbsTemplate.bind({});

WithBreadcrumbs.args = {
  children: <div>Some content</div>,
  breadcrumbs: [
    { href: "#", label: "Home" },
    { href: "#", label: "Item" },
  ],
};
