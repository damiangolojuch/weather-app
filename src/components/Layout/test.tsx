import Layout, { LayoutProps } from ".";
import { render } from "@testing-library/react";

jest.mock("components/Background", () => {
  return {
    __esModule: true,
    default: () => <div data-component="Background" />,
  };
});

jest.mock("antd", () => {
  const ant = jest.requireActual("antd");
  const Layout = (props: any) => (
    <div data-component="Layout">{props.children}</div>
  );
  Layout.Header = (props: any) => (
    <div data-component="Header">{props.children}</div>
  );
  Layout.Content = (props: any) => (
    <div data-component="Content">{props.children}</div>
  );
  Layout.Footer = (props: any) => (
    <div data-component="Footer">{props.children}</div>
  );

  return {
    __esModule: true,
    ...ant,
    Layout,
  };
});

const setup = (customProps: Partial<LayoutProps> = {}) => {
  const defaultProps: LayoutProps = {
    ...customProps,
  };

  return render(<Layout {...defaultProps} />);
};

describe("<Layout />", () => {
  it("should render layout component with Background and Footer component", () => {
    const { container } = setup();

    expect(container).toMatchSnapshot();
  });

  it("should render passed header element", () => {
    const header = <h1>Custom header</h1>;
    const { container } = setup({ header });

    expect(container.getElementsByTagName("h1").item(0)).toHaveTextContent(
      "Custom header"
    );
  });

  it("should render passed children", () => {
    const header = <section>chilren to render</section>;
    const { container } = setup({ header });

    expect(container.getElementsByTagName("section").item(0)).toHaveTextContent(
      "chilren to render"
    );
  });
});
