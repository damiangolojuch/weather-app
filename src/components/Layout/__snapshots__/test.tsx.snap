// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Layout /> should render layout component with Background and Footer component 1`] = `
<div>
  <div
    data-component="Layout"
  >
    <div
      data-component="Background"
    />
    <div
      data-component="Content"
    >
      <div
        class="sc-jSMfEi jUXIxF"
      />
    </div>
    <div
      data-component="Footer"
    >
      Â© Aplikacja pogodowa 2022
    </div>
  </div>
</div>
`;
